openwebwork/webwork2
# Achievement Upgrades

This pull requests gives several upgrades to achievements. 
 1.  Achievements can now be used with gateway and just-in-time sets.  There is an option when you edit an achievement to select which sets that particular achievement is valid for. 
 2.  Achievements are now ordered according to a "Number" field.  Its not quite as nice as drag and drop numbering, but its better than being ordered according to the achievement_id field which couldn't be edited.  
 3.  There are now some achievement items for Gateway Tests. 
 4.  Other minor changes including: 
  -  Using Text::CSV to deal with importing and exporting achievements. 
  -  Adding maketext calls to achievement items. 

To test:  
 1.  Set up a course with achievements already enabled, imported and assigned.  
 2.  Pull this branch and then run a db update on the course.  (The branch adds two new columns to the Achievements table.)  
 3.  Go to the course which had achievements.  The achievements will be missing the number column, but they should still have the same order and if you edit them they should all have the "homework" set type enabled.  
 3.  Delete the achievements and import the default achievements.  Check that achievements are imported from the older style file correctly.  
 4.  Set up a new course copying from the latest version of modelCourse
 5.  Import the default achievements from the new course.  They should all have numbers and should  be assigned to appropriate set types if you use "edit all sets".  
 6.  Check that you can edit achievements and that all of the fields work correctly.  
 6.  Create an achievement with an evaluator that just contains `return 1;`  Test that you can earn this achievement in all three homework types when that type is enabled for the achievement.  
  -  You can unassign achievements you have earned by clicking in the Edit Users column of Achievement Editors. 
 7.  Create an achievement with the evaluator 

  ```
$globalData->{AddNewTestGW} = 1;
$globalData->{ExtendDueDateGW} = 1;
$globalData->{RessurectGW} = 1;
return 1;
  ```
  And earn it.  Check that the three new items work.  (Make sure you have items enabled.)
  -  One item gives you an extra version of a gateway per time period. 
  -  One item extends the due date of the gateway by 24 hours. 
  -  One item reopens a gateway.
